### DISCUSSION TITLE
Are you going to provide a chart/squema where we can see the class structure?

### DISCUSSION THREAD

**Discussion by soymonitus** [SynthstromAudible/DelugeFirmware#21] [Q&A]
Right now looking at the code feels really chaotic and it is like trying to decipher a hieroglyph. It would be really cool if we could have a diagram that for example says (graphically) the Song class contains n Clip instances, with explanations of what each class represents and which is its function. After a look at the code, the initial comment of each file is the typical license agreement text, but I miss some big comment that explains what that class does or represents in the whole application.

**Comment by dctucker** [comment]
I agree that adding doc comments to the top of files and functions would improve the experience for all of us who are new to the codebase. I've taken the liberty of opening [a pull request](https://github.com/SynthstromAudible/DelugeFirmware/pull/20) that adds a basic Doxygen configuration that can be used to publish class diagrams. As comments are added to the codebase, this documentation can be automatically updated, providing additional context.

**Comment by weavermedia** [comment]
Also @jamiefaye mentioned an open source app called Sourcetrail that shows class hierarchy diagrams. It runs on Windows, macOS and linux. It might help. https://github.com/CoatiSoftware/Sourcetrail

**Comment by designerfuzzi** [comment]
apart from visualisers of structures/classes aka Doxygen and co, it makes also sense to declare where such have to be written and in what exact format. As seen all `.cpp` files in the codebase have an according header `.h` file where inline documentation usually belongs in/to.

just an example
```cpp
///Myclass to handle all xyz
class Myclass : InheritedClass {
    ///doStuff does stuff ...
    ///@param size Size of something
    ///@param length length of something
    doStuff(uint32 size, uint16 length); 
}
```

Why the example? Because `///` tripple marking means such documentation is compatible with language servers behind your favourite IDE and will help you coding, not only document stuff. 
That is `// comment` is a comment - not a documentation even if Doxygen knows what to do with it. Doxygen is an external tool to collect written text and classnames etc from within the lines of code but also allows to write ugly documentation which is actually incompatible to language servers (LSP).
